src/main/java/io/quarkus/search/app/util/GitCloneDirectory.java

```diff
@@ -190,13 +190,43 @@ public ObjectId currentSourcesLatestHash() {
      *         or an empty optional otherwise (no submodules)
      */
     public Optional<ObjectId> currentUpstreamSubmoduleSourcesHash() {
+        // We want to get the hash from `git submodule status upstream`, but not from the current state of the branch we are in,
+        //  but rather from a state that we've fetched. As there seems to be no simple way of getting that info out... we will:
+        //  1. Get the name of a current branch we are in (so we can return to it after we are done)
+        //  2. Checkout in detached state the rev we are interested in.
+        //  3. Get the submodule status
+        //  4. Return back to from a detached state to the branch we were initially in.
+        String currentBranch = null;
         try {
-            Map<String, SubmoduleStatus> statusMap = git.submoduleStatus().call();
-            return Optional.ofNullable(statusMap.get("upstream")).map(SubmoduleStatus::getIndexId);
+            currentBranch = git.getRepository().getBranch();
+            try {
+                checkout(currentSourcesLatestHash().name());
+                Map<String, SubmoduleStatus> statusMap = git.submoduleStatus().addPath("upstream").call();
+                return Optional.ofNullable(statusMap.get("upstream")).map(SubmoduleStatus::getIndexId);
+            } catch (GitAPIException e) {
+                throw new RuntimeException(
+                        "Failed to list git submodule status for a repository: " + git.getRepository()
+                                .getRemoteNames() + ". " + e.getMessage(),
+                        e);
+            } finally {
+                checkout(currentBranch);
+            }
+        } catch (IOException e) {
+            Log.error("Wasn't able to correctly determine the current branch: " + e.getMessage(), e);
         } catch (GitAPIException e) {
-            throw new RuntimeException("Failed to list git submodule status for a repository: " + git.getRepository()
-                    .getRemoteNames() + ". " + e.getMessage(), e);
+            Log.error("Wasn't able to correctly switch back to '" + currentBranch + "'" + e.getMessage(), e);
```

Comment:
Shouldn't these be re-thrown? Or does it cause problems with tests?

Reply:
I thought about it, and my idea was that this happens, give the app another chance to try and work with the latest commit from the `main` repository. But then there shouldn't be any obvious reasons for git not to be able to checkout that branch 🤷🏻‍♂️ 😄, so I've changed it to re-throw.